src/components/external-apis/routes.ts

```
@@ -0,0 +1,130 @@
+import { ActionApplicationData } from '../../types'
+import functions from '../actions/evaluatorFunctions'
+import { errorMessage } from '../utilityFunctions'
+
+interface ExternalApiRequest {
+  name: string
+  route: string
+}
+
+const apiConfigs: ExternalApiConfigs = config?.externalApiConfigs ?? {}
+
+export const routeAccessExternalApi = async (
+  request: FastifyRequest & { auth?: { userId?: number; orgId?: number } },
+  reply: FastifyReply
+) => {
+  const { name, route } = request.params as ExternalApiRequest
+
+  const { baseUrl, routes, authentication } = apiConfigs?.[name]
+  if (!baseUrl) {
+    throw new Error('No baseUrl defined for API: ' + name)
+  }
+
+  const routeConfig = routes?.[route]
+  if (!routeConfig) {
+    throw new Error('No config definition for Route: ' + route)
+  }
+
+  const {
+    method,
+    url,
+    permissions,
+    queryParams,
+    allowedClientQueries,
+    returnProperty,
+    additionalAxiosProperties,
+    validationExpression,
+  } = routeConfig
+
+  if (permissions) {
+    const { permissionNames } = await getPermissionNamesFromJWT(request)
+    const validPermissions = permissions.filter((permission) =>
+      permissionNames.includes(permission)
+    )
+
+    if (validPermissions.length === 0) {
+      reply.status(403)
+      return reply.send('Not authorized to view this resource')
+    }
+  }
+
+  // Construct data object for subsequent expression evaluator
+  const { userId, orgId } = request.auth as { userId?: number; orgId?: number }
+  const { user } = await getUserInfo({ userId, orgId })
+  const evaluatorData: {
+    user: typeof user
+    applicationData?: ActionApplicationData
+    functions: typeof functions
+  } = { user, functions }
+
+  // ApplicationData only available if an applicationId is provided as a query
+  // parameter
+  const { applicationId } = request.query as { applicationId?: number }
+  if (applicationId) evaluatorData.applicationData = await getApplicationData({ applicationId })
```

Do we need to check user permission to application ?

---

Yeah, cool. Done. Easiest way to make sure the front-end user can access a certain application, though, was to just make an attempt to fetch the application with user's JWT. If no result, then no go.

---

I've made it so that if the application doesn't pass permission test, it just silently ignores applicationData and carries on with the request. You think it'd be better to return error?

---

Hmm, well technically user should not be on the application if they don't have the permission right ? So either way is fine i guess

---

Yeah, and if the query returns valid content, then the applicationData wasn't needed. And if the applicationData is needed, then they'll get an error anyway if it can't be evaluated. :)